R/variable-as-methods.R

```diff
@@ -0,0 +1,143 @@
+#' categorical variable, the format that the date-time is formatted in (default: 
+#' \code{"\%Y-\%m-\%d \%H:\%M:\%S"}); for `as.Text` and `as.Categorical`, the 
+#' format that the datetime should be formatted as (default: 
+#' \code{"\%Y-\%m-\%d \%H:\%M:\%S"}).
+#' @param resolution for `as.Datetime`, when the variable in `x` is a numeric 
+#' variable, the resolution of the number (e.g. `"ms"` for milliseconds, `"s"` 
+#' for seconds, etc. link)
+#' @param ... additional arguments for `as.character` and `as.numeric`, ignored when used with 
+#' Crunch variables 
+#'
+#' @return a `VariableDefinition` to be used as the derivation 
+#'
+#' @examples
+#' \dontrun{
+#' # ds$v1 is of type Text
+#' is.Text(ds$v1)
+#' # [1] TRUE
+#' 
+#' # that has strings of numbers
+#' as.vector(ds$v1)
+#' # [1] "32"   "8"    "4096" "1024"
+#' 
+#' # convert this to a numeric variable with the alias `v1_numeric`
+#' ds$v1_numeric <- as.Numeric(ds$v1)
+#' 
+#' # the values are the same, but are now numerics and the type is Numeric
+#' as.vector(ds$v1_numeric)
+#' # [1]   32    8 4096 1024
+#' is.Numeric(ds$v1_numeric)
+#' # [1] TRUE
+#' 
+#' # this new variable is derived, so if new data is appended or streamed, the
+#' # new rows of data will be updated.
+#' is.derived(ds$v1_numeric)
+#' # [1] TRUE
+#' }
+#' @name variable-as-methods
+#' @aliases as.Categorical as.Datetime as.Numeric as.Text
+NULL
+
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Numeric", "CrunchVariable", function (x) {
+    haltIfArray(x, callingFunc = "as.Numeric()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        return(zfuncExpr("datetime_to_numeric", x))
+    }
+    
+    return(zfuncExpr("cast", x, "numeric"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Text", "CrunchVariable",
+          function (x, format = "%Y-%m-%d %H:%M:%S") {
+    haltIfArray(x, callingFunc = "as.Text()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        return(zfuncExpr("format_datetime", x, list(value = format)))
+    }
+    
+    return(zfuncExpr("cast", x, "text"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Categorical", "CrunchVariable",
+          function (x, format = "%Y-%m-%d %H:%M:%S") {
+    haltIfArray(x, callingFunc = "as.Categorical()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        # no direct datetime to cat ZCL function, but can just cast the string
+        return(zfuncExpr("cast",
+                         zfunc("format_datetime", x, list(value = format)),
+                         "categorical"))
+    }
+    
+    return(zfuncExpr("cast", x, "categorical"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Datetime", "CrunchVariable", function (x, format, resolution) {
+    haltIfArray(x, callingFunc = "as.Datetime()")
+    
+    # datetimes are special, so each source type must be determined
+    if (is.Numeric(x)) {
+        # TODO: validate resolution
```

Comment:
Might be worth adding a little to this comment explaining the validation you wanted to do.